### PR TITLE
macOS Support

### PR DESCRIPTION
**Pull request by ThadHouse**
Basically restarted the macOS datapath off of the linux datapath rather then the windows datapath. Theyre a lot more similar in how they work, so makes sense to handle it this way.

### CODE REVIEW

**Comment by ThadHouse** [comment]
This also includes all the changes from main, so we might want to merge those in manually, and then rebase this PR.

**Review by nibanks — APPROVED** [review]

File: src/platform/platform_posix.c [hunk 1]
```diff
@@ -776,8 +854,20 @@ CxPlatCurThreadID(
...
+
     CXPLAT_STATIC_ASSERT(sizeof(pid_t) <= sizeof(uint32_t), "PID size exceeds the expected size");
     return syscall(__NR_gettid);
+
+#elif defined(CX_PLATFORM_DARWIN)
+
+    uint64_t tid;
+    pthread_threadid_np(NULL, &tid);
+    // XXX: check for failure
+    return (uint32_t)tid;
```
Comment (wfurt) [inline src/platform/platform_posix.c:868]: This may be problematic when it wraps IMHO. Did you consider changing CxPlatCurThreadID() to return uint64_t? That would cover the Linux as well.

File: src/platform/platform_posix.c [hunk 2]
```diff
@@ -514,7 +516,11 @@ CxPlatGetAbsoluteTime(
 
     CxPlatZeroMemory(Time, sizeof(struct timespec));
 
+#if defined(CX_PLATFORM_LINUX)
     ErrorCode = clock_gettime(CLOCK_MONOTONIC, Time);
+#elif defined(CX_PLATFORM_DARWIN)
```
Comment (wfurt) [inline src/platform/platform_posix.c:521]: I'm not sure why you need this. clock_gettime() should work on macOS just fine AFAIK.

File: src/platform/platform_posix.c (hunk 1 above)
Comment (ThadHouse) [reply src/platform/platform_posix.c:868]: I will check these out tomorrow morning. The one who originally did the mac port did this, so I'll double check everything tomorrow. I'll check the other comment, and any other ones found as well.

File: src/platform/platform_posix.c (hunk 2 above)
Comment (Maximus-) [reply src/platform/platform_posix.c:521]: If I recall correctly, the issue I was experiencing was that clock_gettime was actually *not* monotonic (Seems I wasn't the only one: https://github.com/mobile-shell/mosh/pull/1116), which was breaking some of the pthread event waiters. I feel like my initial PR used mach_absolute_time...

Comment (wfurt) [reply src/platform/platform_posix.c:521]: I would add note there in that case as it is not obvious. .NET runtime uses clock_gettime AFAIK and I'm not aware of any issue. It may not use assert or use that output in the same way.

File: src/platform/platform_posix.c (hunk 1 above)
Comment (wfurt) [reply src/platform/platform_posix.c:868]: maybe add same assert as above. So if it wraps it would be visible. I don't know how this is consumed and probably does not need to addressed right now.

**Review by Anas198900 — APPROVED** [review]

File: src/platform/platform_posix.c [hunk 3]
```diff
@@ -17,6 +17,11 @@
 #include "quic_platform.h"
 #include "quic_platform_dispatch.h"
 #include "quic_trace.h"
+#ifdef CX_PLATFORM_LINUX
+#include <unistd.h>
+#include <sys/types.h>
+#else
+#endif
```
Comment (nibanks) [inline src/platform/platform_posix.c:24]: Is this `#else` necessary?

File: src/platform/platform_posix.c [hunk 4]
```diff
@@ -519,11 +524,14 @@ CxPlatGetAbsoluteTime(
...
+    //
+    // timespec_get is used on darwin, as CLOCK_MONOTONIC isn't actually
+    // monotonic according to our tests.
+    //
+    ErrorCode = timespec_get(Time, TIME_UTC);
 #endif // CX_PLATFORM_DARWIN
 
-    CXPLAT_DBG_ASSERT(ErrorCode == 0);
-    UNREFERENCED_PARAMETER(ErrorCode);
+    CXPLAT_DBG_ASSERT(ErrorCode != 0);
```
Comment (nibanks) [inline src/platform/platform_posix.c:534]: You changed this assert to be the opposite of what it was? Is that not going to break Linux?

File: src/manifest/clog.sidecar [hunk 5]
```diff
@@ -10436,6 +10444,7 @@
         "MacroConfiguration": {
           "linux": "lttng_plus",
           "stubs": "stubs",
+          "macos": "STDOUT",
```
Comment (nibanks) [inline src/manifest/clog.sidecar:10447]: I'm not convinced we practically want to do this.... the spew just for tests will be massive.

File: CMakeLists.txt [hunk 6]
```diff
@@ -336,10 +336,17 @@ else()
...
 
     if(QUIC_ENABLE_LOGGING)
-        message(STATUS "Configuring for LTTng tracing")
-        set(CMAKE_CLOG_CONFIG_PROFILE linux)
-        list(APPEND QUIC_COMMON_DEFINES QUIC_CLOG)
-        include(FindLTTngUST)
+        if (APPLE)
+            message(STATUS "Configuring for macos tracing")
+            set(CMAKE_CLOG_CONFIG_PROFILE macos)
+            #set(CMAKE_CLOG_CONFIG_PROFILE stubs)
```
Comment (nibanks) [inline CMakeLists.txt:342]: Might be better to default to `stubs` instead of `macos` if we are going to use STDOUT for `macos`.

File: CMakeLists.txt [hunk 7]
```diff
@@ -357,7 +364,7 @@ else()
 
     if(QUIC_ENABLE_SANITIZERS)
         message(STATUS "Configuring sanitizers")
-        list(APPEND QUIC_COMMON_FLAGS -fsanitize=address,leak,undefined -fsanitize-address-use-after-scope -Og -ggdb3 -fno-omit-frame-pointer -fno-optimize-sibling-calls)
+        list(APPEND QUIC_COMMON_FLAGS -fsanitize=address,leak,undefined -fsanitize-address-use-after-scope -O0 -ggdb3 -fno-omit-frame-pointer -fno-optimize-sibling-calls)
```
Comment (nibanks) [inline CMakeLists.txt:367]: Why this change?

File: src/platform/platform_posix.c (hunk 4 above)
Comment (ThadHouse) [reply src/platform/platform_posix.c]: Realized that so just fixed it.

File: CMakeLists.txt [hunk 8]
```diff
@@ -478,6 +485,10 @@ if(QUIC_TLS STREQUAL "openssl")
                 linux-armv4 -DL_ENDIAN
                 --cross-compile-prefix=${GNU_MACHINE}${FLOAT_ABI_SUFFIX}-)
             list(APPEND OPENSSL_CONFIG_FLAGS -latomic)
+        elseif(CX_PLATFORM STREQUAL "darwin")
+            # need to build with Apple's compiler
+            # also, "no-shared" breaks the build for master-alpha4-quic-support
```
Comment (nibanks) [inline CMakeLists.txt:490]: Is this comment still valid?

File: CMakeLists.txt (hunk 7 above)
Comment (ThadHouse) [reply CMakeLists.txt:367]: I'm not sure, Max did it.

File: src/core/crypto.c [hunk 9]
```diff
@@ -599,6 +599,7 @@ QuicCryptoWriteOneFrame(
     CXPLAT_DBG_ASSERT(CryptoOffset >= (Crypto->TlsState.BufferTotalLength - Crypto->TlsState.BufferLength));
 
     QUIC_CONNECTION* Connection = QuicCryptoGetConnection(Crypto);
+
```
Comment (nibanks) [inline src/core/crypto.c:602]: nit: remove this extra line

File: CMakeLists.txt (hunk 8 above)
Comment (ThadHouse) [reply CMakeLists.txt:490]: No. Fixed

File: src/inc/quic_platform.h [hunk 10]
```diff
@@ -147,9 +147,9 @@ DEFINE_ENUM_FLAG_OPERATORS(CXPLAT_THREAD_FLAGS);
 #elif _WIN32
 #define CX_PLATFORM_TYPE 2
 #include <quic_platform_winuser.h>
-#elif CX_PLATFORM_LINUX
+#elif CX_PLATFORM_LINUX || CX_PLATFORM_DARWIN
 #define CX_PLATFORM_TYPE 3
-#include <quic_platform_linux.h>
+#include <quic_platform_posix.h>
 #else
```
Comment (nibanks) [inline src/inc/quic_platform.h:153]: What do you think about defining different `CX_PLATFORM_TYPE` values here instead, but both still including *posix.h?

File: src/platform/platform_posix.c [hunk 11]
```diff
@@ -771,13 +858,26 @@ CxPlatThreadWait(
...
+    CXPLAT_STATIC_ASSERT(sizeof(pid_t) <= sizeof(CXPLAT_THREAD_ID), "PID size exceeds the expected size");
+    return gettid();
+
+#elif defined(CX_PLATFORM_DARWIN)
+
+    uint64_t Tid;
+    int Res = pthread_threadid_np(NULL, &Tid);
+    CXPLAT_DBG_ASSERT(Res == 0);
+    CXPLAT_DBG_ASSERT(Tid <= UINT32_MAX);
+    return (uint32_t)Tid;
```
Comment (nibanks) [inline src/platform/platform_posix.c:877]:
```suggestion
    return (CXPLAT_THREAD_ID)Tid;
```

File: src/platform/platform_posix.c [hunk 12]
```diff
@@ -771,13 +858,26 @@ CxPlatThreadWait(
...
     void
     )
 {
-    CXPLAT_STATIC_ASSERT(sizeof(pid_t) <= sizeof(uint32_t), "PID size exceeds the expected size");
-    return syscall(__NR_gettid);
+
+#if defined(CX_PLATFORM_LINUX)
+
+    CXPLAT_STATIC_ASSERT(sizeof(pid_t) <= sizeof(CXPLAT_THREAD_ID), "PID size exceeds the expected size");
+    return gettid();
```
Comment (nibanks) [inline src/platform/platform_posix.c:869]: Why change this from what it was?

Comment (ThadHouse) [reply src/platform/platform_posix.c:869]: Because we shouldn't be using syscalls directly, and especially not the NR versions of the syscalls. gettid returns the exact same value, but actually in a posix compatible way.

File: src/inc/quic_platform.h [hunk 13]
```diff
@@ -149,7 +149,10 @@ DEFINE_ENUM_FLAG_OPERATORS(CXPLAT_THREAD_FLAGS);
 #include <quic_platform_winuser.h>
 #elif CX_PLATFORM_LINUX
 #define CX_PLATFORM_TYPE 3
-#include <quic_platform_linux.h>
+#include <quic_platform_posix.h>
+elif CX_PLATFORM_DARWIN
```
Comment (nibanks) [inline src/inc/quic_platform.h:153]: `#elif`

File: CMakeLists.txt (hunk 7 above)
Comment (Maximus-) [reply CMakeLists.txt]: I changed it to -O0 because -Og is almost *never* what you want. -Og is "optimize for debuggability, in production", but closer to -O1, so real optimizations do happen which make debugging harder. 

Also nit: on macOS, it's -glldb for optimize debugging for lldb =) since gdb on macOS is mostly broken.
